egui_glow/src/misc_util.rs

```diff
@@ -0,0 +1,186 @@
+            glow::TEXTURE_WRAP_T,
+            glow::CLAMP_TO_EDGE as i32,
+        );
+        if is_webgl_1 {
+            let format = if srgb_support {
+                glow::SRGB_ALPHA
+            } else {
+                glow::RGBA
+            };
+            gl.tex_image_2d(
+                glow::TEXTURE_2D,
+                0,
+                format as i32,
+                w as i32,
+                h as i32,
+                0,
+                format,
+                glow::UNSIGNED_BYTE,
+                Some(data),
+            );
+        } else {
+            gl.tex_storage_2d(glow::TEXTURE_2D, 1, glow::SRGB8_ALPHA8, w as i32, h as i32);
+            gl.tex_sub_image_2d(
+                glow::TEXTURE_2D,
+                0,
+                0,
+                0,
+                w as i32,
+                h as i32,
+                glow::RGBA,
+                glow::UNSIGNED_BYTE,
+                glow::PixelUnpackData::Slice(data),
+            );
+        }
+        assert_eq!(gl.get_error(), glow::NO_ERROR, "OpenGL error occurred!");
+        tex
+    }
+}
+
+pub(crate) unsafe fn as_u8_slice<T>(s: &[T]) -> &[u8] {
+    std::slice::from_raw_parts(s.as_ptr().cast::<u8>(), s.len() * std::mem::size_of::<T>())
+}
+
+#[cfg(target_arch = "wasm32")]
+pub(crate) fn glow_debug_print(s: impl Into<JsValue>) {
+    web_sys::console::log_1(&s.into());
+}
+#[cfg(not(target_arch = "wasm32"))]
+pub(crate) fn glow_debug_print(s: impl std::fmt::Display) {
+    println!("{}", s);
+}
+
+pub(crate) fn compile_shader(
+    gl: &glow::Context,
+    shader_type: u32,
+    source: &str,
+) -> Result<glow::Shader, String> {
+    let shader = unsafe { gl.create_shader(shader_type) }?;
+    unsafe {
+        gl.shader_source(shader, source);
+    }
+    unsafe {
+        gl.compile_shader(shader);
+    }
+
+    if unsafe { gl.get_shader_compile_status(shader) } {
+        Ok(shader)
+    } else {
+        Err(unsafe { gl.get_shader_info_log(shader) })
+    }
+}
+
+pub(crate) fn link_program<'a, T: IntoIterator<Item = &'a glow::Shader>>(
+    gl: &glow::Context,
+    shaders: T,
+) -> Result<glow::Program, String> {
+    let program = unsafe { gl.create_program() }?;
+    unsafe {
+        for shader in shaders {
+            gl.attach_shader(program, *shader);
+        }
+    }
+    unsafe {
+        gl.link_program(program);
+    }
+
+    if unsafe { gl.get_program_link_status(program) } {
+        Ok(program)
+    } else {
+        Err(unsafe { gl.get_program_info_log(program) })
+    }
+}
+///Wrapper around Emulated VAO and GL's VAO
+pub(crate) enum VAO {
+    Emulated(crate::vao_emulate::EmulatedVao),
+    Native(crate::glow::VertexArray),
+}
+
+impl VAO {
+    pub(crate) unsafe fn new(gl: &glow::Context, is_native_vao: bool) -> Self {
+        if is_native_vao {
+            Self::Native(gl.create_vertex_array().unwrap())
+        } else {
+            Self::Emulated(crate::vao_emulate::EmulatedVao::new())
+        }
+    }
```

Comment:
This is nitpicky, but I think it would make sense to offer two constructors here instead of one.

```suggestion
    pub(crate) unsafe fn native(gl: &glow::Context) -> Self {
        Self::Native(gl.create_vertex_array().unwrap())
    }
    pub(crate) fn emulated() -> Self {
        Self::Emulated(crate::vao_emulate::EmulatedVao::new())
    }
```

Also, wouldn't the API be nicer if the constructor captured the `&glow::Context` here so it doesn't need to be passed to the other methods? That might also help reduce some confusion about why `VAO::bind_vertex_array` is unsafe, but the others are not?

Reply:
Good point about two constructors 👍 

Capturing `&glow::Context` leads to a lot of ugly lifetime issues. Passing it in by reference is also more explicit, i.e. easier to see what is actually doing graphics calls. This was discussed in the original glow PR.

Reply:
Ok  i will split constructor.
but is there good way to switch emulated mode?

Reply:
> Capturing `&glow::Context` leads to a lot of ugly lifetime issues. Passing it in by reference is also more explicit, i.e. easier to see what is actually doing graphics calls. This was discussed in the original glow PR.

Ok! Good to know, I didn't follow that conversation closely.

> Ok i will split constructor.
> but is there good way to switch emulated mode?

This question implies there is a way to switch the emulation mode with just one constructor? I don't believe that is the case before or after the suggested patch.

Reply:
I thought this procedure good enough to switch emulated mode .
Is there any suggestion?
![vao_switch_procedure](https://user-images.githubusercontent.com/48007646/140046237-6f72c1a2-fec4-49cf-81dc-41a9d37d08cd.png)